pandas/core/indexes/base.py

```diff
@@ -3567,16 +3566,26 @@ def _join_multi(self, other, how, return_indexers=True):
     def _join_non_unique(self, other, how="left", return_indexers=False):
         from pandas.core.reshape.merge import _get_join_indexers
 
+        # We only get here if dtypes match
+        assert self.dtype == other.dtype
+
+        if is_extension_array_dtype(self.dtype):
```

Comment:
I would make

extract_values_for_argsort(ArrayLike) -> np.array

which does this for a single array and just call it here (and below); locate in pandas/core/consruction near extract_array

Reply:
there are a couple more places in Index where I _think_ this same pattern can be used.  once that is verified, i was thinking of making this a method in Index/ExtensionIndex

Reply:
gentle ping.  ive got a branch that does this for the remaining places in indexes.base and does implement this as `_get_engine_target`